### PR TITLE
Lab1

### LINKED ISSUE #413: Build lab 1 assignment code for Fibonacci and Negafibonacci series

### PR DESCRIPTION
**Pull request by nicko-lee**
Fixes #413
Review of colleague's PR #498
Implement fibonacci and write tests.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
```
Comment (joshcarp) [inline .vscode/launch.json]: this shouldn't be committed

File: 01_fib/nickolee/fibonacci.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+		return result[1:]
+	}
+
+	return []int{}
+}
+
+func main() {
+	// using the for-each range loop described here: https://yourbasic.org/golang/for-loop/
+	for _, v := range fib(7) {
+		fmt.Fprintf(out, "%d\n", v)
```
Comment (joshcarp) [inline 01_fib/nickolee/fibonacci.go]: why doesn't this get printed within fib?

File: 01_fib/nickolee/fibonacci_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"testing"
+)
+
+func TestFibonacci(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := strconv.Quote(`1
```
Comment (joshcarp) [inline 01_fib/nickolee/fibonacci_test.go:16]: any reason you're not using an interpreted string literal?

Comment (joshcarp) [reply 01_fib/nickolee/fibonacci_test.go:16]: `"1\n1\n2\n3\n5\n8\n13\n"`

File: 01_fib/nickolee/fibonacci_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	if expected != actual {
+		t.Errorf("TestFibonacci: Unexpected output in main()")
+	}
+}
+
+func TestFibonacciUpperLimit(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+	fib(93)
+	expected := strconv.Quote("Overflow error. Please use an argument less than 92")
```
Comment (joshcarp) [inline 01_fib/nickolee/fibonacci_test.go:35]: any reason why you're using strconv.Quote()?

File: 01_fib/nickolee/fibonacci_test.go (hunk 3 above)
Comment (kasunfdo) [reply 01_fib/nickolee/fibonacci_test.go:16]: @Joshcarp, in wiki section, it is asked to [use raw strings](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#consider-using-raw-strings)

**Review by shuutech — COMMENTED** [review]
01_fib/nickolee/fibonacci.go
Your case takes into account the upper limit of 92 for 64bit int but the code is using int. Would an overflow error occur if n=91? Maybe set a lower limit e.g. 88 and check if you receive the overflow error

**Review by shuutech — CHANGES_REQUESTED** [review]
Your case takes into account the upper limit of 92 for 64bit int but the code is using int. Would an overflow error occur if n=91? Maybe set a lower limit e.g. 88 and check if you receive the overflow error

**Review by juliaogris — CHANGES_REQUESTED** [review]
Hi @shuutech 	- good job thinking of overflows! We have decided we are going to ignore them, but if you want to be perfectionist (we love that ;) ) you could add a check that n is no larger than 92 (be careful though, `int` defaults to 32 bits on some systems e.g. https://goplay.space/)

Please also have a look at [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)
- [Fix your PR description](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#pr-description)
- [Limit the number of commits to one or two](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#limit-the-number-of-commits-to-one-or-two)

**Comment by rokane** [comment]
Hey mate, just took a look at your commit messages and the comments here. I know git can be a bit daunting at times, but take a look at doing a rebase and squashing those `[WIP]` commits into one commit and renaming it so that it is something meaningful. Once you have effectively "rewritten" the git history, you can force push `git push +branch_name` and it will overwrite the history here. I can help you out if need be, feel free to drop by my desk sometime and we can do the rebase together.

Also make sure you do a review of someone else's and include some details in the Pull Request description.

**Comment by shuutech** [comment]
> Hi @shuutech - good job thinking of overflows! We have decided we are going to ignore them, but if you want to be perfectionist (we love that ;) ) you could add a check that n is no larger than 92 (be careful though, `int` defaults to 32 bits on some systems e.g. https://goplay.space/)
> 
> Please also have a look at [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)
> 
> * [Fix your PR description](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#pr-description)
> * [Limit the number of commits to one or two](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#limit-the-number-of-commits-to-one-or-two)

Hi Julia,
The code reviewed had included a check that n is no larger than 92, however there was a comment that this check was relevant for int64. As int was used instead of int64, I just wanted to clarify if the author intended to use int64 or int.

**Review by juliaogris — CHANGES_REQUESTED** [review]
remove binary file `00_hello_world/juliaogris/helloworld` (2MB) and .vscode files please and rebase.

File: 01_fib/nickolee/fibonacci.go (hunk 2 above)
Comment (nicko-lee) [reply 01_fib/nickolee/fibonacci.go]: Fixed

File: .vscode/launch.json (hunk 1 above)
Comment (nicko-lee) [reply .vscode/launch.json]: Fixed

File: 01_fib/nickolee/fibonacci_test.go (hunk 4 above)
Comment (nicko-lee) [reply 01_fib/nickolee/fibonacci_test.go:35]: Followed Julia's example in 00_hello_world/juliaogris/helloworld

**Comment by alanhillgemann** [comment]
@nicko-lee the upper limit of 92 does not apply when executing on a 32bit system unless explicitly declaring as int64.

**Pull request merged**